src/main/scala/is/hail/variant/VariantDataset.scala

```diff
@@ -44,59 +50,93 @@ object VariantDataset {
-                row.getSeq[Row](2).lazyMap { rg =>
-                  new RowGenotype(rg): Genotype
-                }))
-          } else
+                Iterable.empty[Annotation])))
+        else {
+          val rdd = sqlContext.readParquetSorted(parquetFile)
           rdd.map { row =>
             val v = row.getVariant(0)
             (v,
               (if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(row.get(1), vaSignature) else row.get(1),
-                row.getGenotypeStream(v, 2, isDosage): Iterable[Genotype]))
+                row.getSeq[Any](2).lazyMap { g => if (gRequiresConversion) SparkAnnotationImpex.importAnnotation(g, genotypeSignature) else g }
+              ))
           }
+        }
+
+        OrderedRDD(rdd, partitioner)
       }
 
-      val partitioner: OrderedPartitioner[Locus, Variant] =
-        try {
-          val jv = hConf.readFile(dirname + "/partitioner.json.gz")(JsonMethods.parse(_))
-          jv.fromJSON[OrderedPartitioner[Locus, Variant]]
-        } catch {
-          case _: FileNotFoundException =>
-            fatal("missing partitioner.json.gz when loading VDS, create with HailContext.write_partitioning.")
+      new VariantSampleMatrix[Annotation](hc, metadata, localValue, orderedRDD)
+    } else {
+      val orderedRDD = if (skipVariants)
+        OrderedRDD.empty[Locus, Variant, (Annotation, Iterable[Genotype])](sc)
+      else {
+        val rdd = if (skipGenotypes)
+          sqlContext.readParquetSorted(parquetFile, Some(Array("variant", "annotations")))
+            .map(row => (row.getVariant(0),
+              (if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(row.get(1), vaSignature) else row.get(1),
+                Iterable.empty[Genotype])))
+        else {
+          val rdd = sqlContext.readParquetSorted(parquetFile)
+          if (parquetGenotypes)
+            rdd.map { row =>
+              val v = row.getVariant(0)
+              (v,
+                (if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(row.get(1), vaSignature) else row.get(1),
+                  row.getSeq[Row](2).lazyMap { rg =>
+                    new RowGenotype(rg): Genotype
+                  }))
+            } else
+            rdd.map { row =>
+              val v = row.getVariant(0)
+              (v,
+                (if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(row.get(1), vaSignature) else row.get(1),
+                  row.getGenotypeStream(v, 2, isDosage): Iterable[Genotype]))
+            }
         }
 
-      OrderedRDD(rdd, partitioner)
-    }
+        OrderedRDD(rdd, partitioner)
+      }
 
-    new VariantSampleMatrix[Genotype](hc,
-      if (skipGenotypes) metadata.copy(sampleIds = IndexedSeq.empty[String],
-        sampleAnnotations = IndexedSeq.empty[Annotation])
-      else metadata,
-      orderedRDD)
+      new VariantSampleMatrix[Genotype](hc, metadata, localValue, orderedRDD)
+    }
```

Comment:
Times are measured in nanoseconds, so most of this stuff took on the order of milliseconds. Perhaps larger scale tests should be run too?